presto-spark/src/main/java/com/facebook/presto/spark/TaskProcessors.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark;
+
+import org.apache.spark.TaskContext;
+import org.apache.spark.api.java.function.FlatMapFunction2;
+import org.apache.spark.api.java.function.PairFlatMapFunction;
+import org.apache.spark.util.CollectionAccumulator;
+import scala.Tuple2;
+
+import java.util.HashMap;
+import java.util.Iterator;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonMap;
+import static java.util.Collections.unmodifiableMap;
+
+public class TaskProcessors
+{
+    private TaskProcessors() {}
+
+    public static PairFlatMapFunction<byte[], Integer, byte[]> createTaskProcessor(
+            PrestoSparkTaskCompilerFactory taskCompilerFactory,
+            CollectionAccumulator<byte[]> taskStatsCollector)
+    {
+        return (serializedTaskDescriptor) -> {
+            int taskId = TaskContext.get().partitionId();
```

Comment:
In fact, consider using `mapPartitionsWithIndex`: https://spark.apache.org/docs/2.1.1/api/java/org/apache/spark/api/java/JavaRDD.html#mapPartitionsWithIndex(org.apache.spark.api.java.function.Function2,%20boolean)

Unfortunately there is no `mapPartitionsToPairWithIndex` :(

Reply:
We don't need to have the partition id for every value (it will be the same). We need to extract partition id only once.